installTerraInVm.sh

```diff
@@ -1,53 +1,54 @@
+#!/bin/bash
+
 # Check if gnupg is installed
-if which gpg > /dev/null 2>&1; then
-  echo 'gnupg and software-properties-common are already installed'
+if gpg --version; then
+    echo 'gnupg is already installed'
 else
-  echo 'Installing gnupg and software-properties-common...'
-  sudo apt-get update && sudo apt-get install -y gnupg software-properties-common
+    sudo apt-get update
+    echo 'Installing gnupg and software-properties-common...'
+    sudo apt-get install -y gnupg software-properties-common
 fi
 
-# Check if the HashiCorp GPG key is present
-if gpg --no-default-keyring --keyring /usr/share/keyrings/hashicorp-archive-keyring.gpg --fingerprint | grep -q 'HashiCorp'; then
-  echo 'HashiCorp GPG key is already present.'
+# Check for apt-cache status
+if apt-cache search gnupg && apt-cache search software-properties-common; then
+    echo 'Package cache is up to date'
 else
-  echo 'HashiCorp GPG key not found. Adding the key now.'
-  wget -O- https://apt.releases.hashicorp.com/gpg | \
-  gpg --dearmor | \
-  sudo tee /usr/share/keyrings/hashicorp-archive-keyring.gpg > /dev/null
+    echo 'Updating package cache...'
+    sudo apt-get update
 fi
 
-# Verify the key's fingerprint
-if gpg --no-default-keyring --keyring /usr/share/keyrings/hashicorp-archive-keyring.gpg --fingerprint | grep -q 'HashiCorp'; then
-  echo 'HashiCorp GPG key fingerprint verified.'
+# Check for HashiCorp GPG key
+if gpg --no-default-keyring --keyring /usr/share/keyrings/hashicorp-archive-keyring.gpg --fingerprint | grep HashiCorp; then
+    echo 'HashiCorp GPG key is already present and verified.'
 else
-  echo 'Error: HashiCorp GPG key fingerprint verification failed.'
-  exit 1
+    echo 'Adding HashiCorp GPG key...'
+    wget -O- https://apt.releases.hashicorp.com/gpg | \
+    gpg --dearmor | \
+    sudo tee /usr/share/keyrings/hashicorp-archive-keyring.gpg > /dev/null
 fi
 
-# Check if the HashiCorp repository is already present
-if grep -q 'hashicorp.list' /etc/apt/sources.list.d/hashicorp.list; then
-  echo 'HashiCorp repository is already present.'
+# Check for HashiCorp repository
+if test -f /etc/apt/sources.list.d/hashicorp.list; then
+    echo 'HashiCorp repository is already present.'
 else
-  echo 'Adding HashiCorp repository.'
-  echo "deb [signed-by=/usr/share/keyrings/hashicorp-archive-keyring.gpg] \
+    echo 'Adding HashiCorp repository...'
+    echo "deb [signed-by=/usr/share/keyrings/hashicorp-archive-keyring.gpg] \
 https://apt.releases.hashicorp.com $(lsb_release -cs) main" | \
-  sudo tee /etc/apt/sources.list.d/hashicorp.list
+    sudo tee /etc/apt/sources.list.d/hashicorp.list > /dev/null
 fi
 
-# Update package information from HashiCorp
-echo 'Updating package information from HashiCorp repository.'
-sudo apt update
+# Check if apt needs updating after potential repository addition
+if test -f /etc/apt/sources.list.d/hashicorp.list && find /etc/apt/sources.list.d/hashicorp.list -mmin -5; then
```

Comment:
The check for terraform to be in the apt-cache would be good enough you don't need to search in the sources and this doesn't work anyway.

You can just throw another check like:

```
if (apt-cache show terraform)
then
...
else
  apt update
fi
```